dot/state/block.go

```diff
@@ -250,6 +251,39 @@ func (bs *BlockState) GetHashByNumber(num uint) (common.Hash, error) {
 	return common.NewHash(bh), nil
 }
 
+// GetHashesByNumber returns the block hashes with the given number
+func (bs *BlockState) GetHashesByNumber(num uint) ([]common.Hash, error) {
+	block, err := bs.GetBlockByNumber(num)
+	if err != nil {
+		return []common.Hash{}, fmt.Errorf("getting block by number %w", err)
+	}
+	blockHashes := bs.bt.GetAllBlocksAtNumber(block.Header.ParentHash)
+
+	hash, err := bs.bt.GetHashByNumber(num)
+	if err != nil {
+		if !errors.Is(err, blocktree.ErrNumLowerThanRoot) {
+			return blockHashes, fmt.Errorf("failed to get hash from blocktree: %w", err)
+		}
+
+		// if error is ErrNumLowerThanRoot, number has already been finalised, so check db
+		bh, err := bs.db.Get(headerHashKey(uint64(num)))
+		if err != nil {
+			return blockHashes, fmt.Errorf("cannot get block %d: %w", num, err)
+		}
+
+		if !slices.Contains(blockHashes, hash) {
+			blockHashes = append(blockHashes, common.NewHash(bh))
+		}
+
+		return blockHashes, nil
+	}
+
+	if !slices.Contains(blockHashes, hash) {
```

Comment:
Since the function `GetAllBlocksAtNumber` returns all block hashes with the number of the given hash plus one why we need to check if the `blockHashes` contains `hash`?

Reply:
`GetAllBlocksAtNumber` does not return all block hashes, only the ones that are in memory.

`bs.GetBlockByNumber(num)` looks into both disk and memory to give you a block.
`bs.bt.GetAllBlocksAtNumber` looks only inside memory.

If hash was read from memory, blockHashes would have it.
If it was read from disk, that means it was pruned from memory so blockHashes won't have it.